README
NAME
    Crypt::NaCl::Sodium - NaCl compatible modern, easy-to-use library for
    encryption, decryption, signatures, password hashing and more

VERSION
    version 1.0.7.0

SYNOPSIS
        use Crypt::NaCl::Sodium qw( :utils );

        my $crypto = Crypt::NaCl::Sodium->new();

        ##########################
        ## Secret-key cryptography

        # Secret-key authenticated encryption (XSalsa20/Poly1305 MAC)
        my $crypto_secretbox = $crypto->secretbox();

        # Secret-key message authentication (HMAC-SHA256, HMAC-SHA512, HMAC-SHA512/256 )
        my $crypto_auth = $crypto->auth();

        # Authenticated Encryption with Additional Data (ChaCha20/Poly1305 MAC, AES256-GCM)
        my $crypto_aead = $crypto->aead();

        ##########################
        ## Public-key cryptography

        # Public-key authenticated encryption (Curve25519/XSalsa20/Poly1305 MAC)
        my $crypto_box = $crypto->box();

        # Public-key signatures (Ed25519)
        my $crypto_sign = $crypto->sign();

        ##########################
        ## Hashing

        # Generic hashing (Blake2b)
        my $crypto_generichash = $crypto->generichash();

        # Short-input hashing (SipHash-2-4)
        my $crypto_shorthash = $crypto->shorthash();

        ##########################
        ## Password hashing (yescrypt)

        my $crypto_pwhash = $crypto->pwhash();

        ##########################
        ## Advanced

        # SHA-2 (SHA-256, SHA-512)
        my $crypto_hash = $crypto->hash();

        # One-time authentication (Poly1305)
        my $crypto_onetimeauth = $crypto->onetimeauth();

        # Diffie-Hellman (Curve25519)
        my $crypto_scalarmult = $crypto->scalarmult();

        # Stream ciphers (XSalsa20, ChaCha20, Salsa20, AES-128-CTR)
        my $crypto_stream = $crypto->stream();

        ##########################
        ## Utilities

        # convert binary data to hexadecimal
        my $hex = bin2hex($bin);

        # convert hexadecimal to binary
        my $bin = hex2bin($hex);

        # constant time comparision of strings
        memcmp($a, $b, $length ) or die '$a ne $b';

        # constant time comparision of large numbers
        compare($x, $y, $length ) == -1 and print '$x < $y';

        # overwrite with null bytes
        memzero($a, $b, ...);

        # generate random number
        my $num = random_number($upper_bound);

        # generate random bytes
        my $bytes = random_bytes($count);

        ##########################
        ## Guarded data storage

        my $locker = Data::BytesLocker->new($password);
        ...
        $locker->unlock();
        print $locker->to_hex();
        $locker->lock();

DESCRIPTION
    Crypt::NaCl::Sodium provides bindings to libsodium - NaCl compatible
    modern, easy-to-use library for encryption, decryption, signatures,
    password hashing and more.

    It is a portable, cross-compilable, installable, packageable fork of
    NaCl <http://nacl.cr.yp.to/>, with a compatible API, and an extended API
    to improve usability even further.

    Its goal is to provide all of the core operations needed to build
    higher-level cryptographic tools.

    The design choices emphasize security, and "magic constants" have clear
    rationales.

    And despite the emphasis on high security, primitives are faster
    across-the-board than most implementations of the NIST standards.

    Crypt::NaCl::Sodium uses Alien::Sodium that tracks the most current
    releases of libsodium.

METHODS
  new
        my $crypto = Crypt::NaCl::Sodium->new();

    Returns a proxy object for methods provided below.

  secretbox
        # Secret-key authenticated encryption (XSalsa20/Poly1305 MAC)
        my $crypto_secretbox = Crypt::NaCl::Sodium->secretbox();

    Read Crypt::NaCl::Sodium::secretbox for more details.

  auth
        # Secret-key authentication (HMAC-SHA512/256 and advanced usage of HMAC-SHA-2)
        my $crypto_auth = Crypt::NaCl::Sodium->auth();

    Read Crypt::NaCl::Sodium::auth for more details.

  aead
        # Authenticated Encryption with Additional Data (ChaCha20/Poly1305 MAC, AES256-GCM)
        my $crypto_aead = Crypt::NaCl::Sodium->aead();

    Read Crypt::NaCl::Sodium::aead for more details.

  box
        # Public-key authenticated encryption (Curve25519/XSalsa20/Poly1305 MAC)
        my $crypto_box = Crypt::NaCl::Sodium->box();

    Read Crypt::NaCl::Sodium::box for more details.

  sign
        # Public-key signatures (Ed25519)
        my $crypto_sign = Crypt::NaCl::Sodium->sign();

    Read Crypt::NaCl::Sodium::sign for more details.

  generichash
        # Generic hashing (Blake2b)
        my $crypto_generichash = Crypt::NaCl::Sodium->generichash();

    Read Crypt::NaCl::Sodium::generichash for more details.

  shorthash
        # Short-input hashing (SipHash-2-4)
        my $crypto_shorthash = Crypt::NaCl::Sodium->shorthash();

    Read Crypt::NaCl::Sodium::shorthash for more details.

  pwhash
        # Password hashing (yescrypt)
        my $crypto_pwhash = Crypt::NaCl::Sodium->pwhash();

    Read Crypt::NaCl::Sodium::pwhash for more details.

  hash
        # SHA-2 (SHA-256, SHA-512)
        my $crypto_hash = Crypt::NaCl::Sodium->hash();

    Read Crypt::NaCl::Sodium::hash for more details.

  onetimeauth
        # One-time authentication (Poly1305)
        my $crypto_onetimeauth = Crypt::NaCl::Sodium->onetimeauth();

    Read Crypt::NaCl::Sodium::onetimeauth for more details.

  scalarmult
        # Diffie-Hellman (Curve25519)
        my $crypto_scalarmult = Crypt::NaCl::Sodium->scalarmult();

    Read Crypt::NaCl::Sodium::scalarmult for more details.

  stream
        # Stream ciphers (XSalsa20, ChaCha20, Salsa20, AES-128-CTR)
        my $crypto_stream = Crypt::NaCl::Sodium->stream();

    Read Crypt::NaCl::Sodium::stream for more details.

FUNCTIONS
        use Crypt::NaCl::Sodium qw(:utils);

    Imports all provided functions.

  bin2hex
        my $hex = bin2hex($bin);

    Returns converted $bin into a hexadecimal string.

  hex2bin
        my $hex = "41 : 42 : 43";
        my $bin = hex2bin($hex, ignore => ": ", max_len => 2 );
        print $bin; # AB

    Parses a hexadecimal string $hex and converts it to a byte sequence.

    Optional arguments:

    *   ignore

        A string of characters to skip. For example, the string ": " allows
        columns and spaces to be present at any locations in the hexadecimal
        string. These characters will just be ignored.

        If unset any non-hexadecimal characters are disallowed.

    *   max_len

        The maximum number of bytes to return.

    The parser stops when a non-hexadecimal, non-ignored character is found
    or when "max_len" bytes have been written.

  memcmp
        memcmp($a, $b, $length ) or die "\$a ne \$b for length: $length";

    Compares strings in constant-time. Returns true if they match, false
    otherwise.

    The argument $length is optional if variables are of the same length.
    Otherwise it is required and cannot be greater then the length of the
    shorter of compared variables.

    NOTE: "memcmp" in Data::BytesLocker provides the same functionality.

        $locker->memcmp($b, $length) or die "\$locker ne \$b for length: $length";

  compare
        compare($x, $y, $length ) == -1 and print '$x < $y';

    A constant-time version of "memcmp", useful to compare nonces and
    counters in little-endian format, that plays well with "increment".

    Returns -1 if $x is lower then $y, 0 if $x and $y are identical, or 1 if
    $x is greater then $y. Both $x and $y are assumed to be numbers encoded
    in little-endian format.

    The argument $length is optional if variables are of the same length.
    Otherwise it is required and cannot be greater then the length of the
    shorter of compared variables.

    NOTE: "compare" in Data::BytesLocker provides the same functionality.

        $locker->compare($y, $length) == -1 and print "\$locker < \$y for length: $length";

  memzero
        memzero($a, $b, ...);

    Replaces the value of the provided stringified variables with "null"
    bytes. Length of the zeroed variables is unchanged.

  random_number
        my $num = random_number($upper_bound);

    Returns an unpredictable number between 0 and optional $upper_bound
    (excluded). If $upper_bound is not specified the maximum value is
    0xffffffff (included).

  increment
        increment($nonce, ...);

    NOTE: This function is deprecated and will be removed in next version.
    Please use "increment" in Data::BytesLocker.

    Increments an arbitrary long unsigned number(s) (in place). Function
    runs in constant-time for a given length of arguments and considers them
    to be encoded in little-endian format.

  random_bytes
        my $bytes = random_bytes($num_of_bytes);

    Generates unpredictable sequence of $num_of_bytes bytes.

    The length of the $bytes equals the value of $num_of_bytes.

    Returns Data::BytesLocker object.

VARIABLES
  $Data::BytesLocker::DEFAULT_LOCKED
        use Crypt::NaCl::Sodium;
        $Data::BytesLocker::DEFAULT_LOCKED = 1;

    By default all values returned from the provided methods are unlocked
    Data::BytesLocker objects. If this variable is set to true then the
    returned objects are locked and require calling "unlock" in
    Data::BytesLocker before accessing.

SEE ALSO
    *   Crypt::NaCl::Sodium::secretbox - Secret-key authenticated encryption
        (XSalsa20/Poly1305 MAC)

    *   Crypt::NaCl::Sodium::auth - Secret-key message authentication
        (HMAC-SHA256, HMAC-SHA512, HMAC-SHA512/256 )

    *   Crypt::NaCl::Sodium::aead - Authenticated Encryption with Additional
        Data (ChaCha20/Poly1305 MAC, AES256-GCM)

    *   Crypt::NaCl::Sodium::box - Public-key authenticated encryption
        (Curve25519/XSalsa20/Poly1305 MAC)

    *   Crypt::NaCl::Sodium::sign - Public-key signatures (Ed25519)

    *   Crypt::NaCl::Sodium::generichash - Generic hashing (Blake2b)

    *   Crypt::NaCl::Sodium::shorthash - Short-input hashing (SipHash-2-4)

    *   Crypt::NaCl::Sodium::pwhash - Password hashing (yescrypt)

    *   Crypt::NaCl::Sodium::hash - SHA-2 (SHA-256, SHA-512)

    *   Crypt::NaCl::Sodium::onetimeauth - One-time authentication
        (Poly1305)

    *   Crypt::NaCl::Sodium::scalarmult - Diffie-Hellman (Curve25519)

    *   Crypt::NaCl::Sodium::stream - Stream ciphers (XSalsa20, ChaCha20,
        Salsa20, AES-128-CTR)

    *   Data::BytesLocker - guarded data storage

    *   libsodium <http://jedisct1.gitbooks.io/libsodium> - libsodium

AUTHOR
    Alex J. G. Burzyński <ajgb@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by Alex J. G. Burzyński
    <ajgb@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Crypt::NaCl::Sodium.

{{$NEXT}}

1.0.8.0 2015-12-27 21:45:43
    - Use Alien::Sodium v1.0.8.0

1.0.7.0 2015-12-24 02:44:37
    - New methods added to Data::BytesLocker: is_zero(), increment(), add(), clone().
    - Helper methods memcmp(), compare() added to Data::BytesLocker
    - Require Alien::Sodium v1.0.7.0 (to enable new features)
    - DEPRECATED: increment() function - use Data::BytesLocker's increment method

1.0.6.1 2015-11-25 23:48:38
    - Fix chacha20_ietf_xor_ic on 32bit platforms

1.0.6.0 2015-11-22 22:57:54
    - Verify if AES256-GCM is available and die in related methods if not.
    - increment() accepts variable list of arguments.
    - cleanup documentation.

0.09   2015-11-10 02:15:01
    - Data::BytesLocker bytes() method added.
    - box, secretbox and aead verify the minimum length of ciphertext.
    - Require Alien::Sodium v1.0.6.0 (to enable new features)
    - Added following compare and increment utility functions.
    - IETF-compatible version of ChaCha20 stream cipher
    - Added support for AES256-GCM.

0.08   2015-07-17 00:15:13
    - Require Alien::Sodium v1.0.3.1

0.07   2015-07-13 22:37:05
    - Allow to specify library and include directory paths
  if pre-installed libsodium is to be used [gh#1]
    - Minor documentation changes

0.06   2015-05-20 22:37:45
    - Build changes to locate libsodium.a if pkg-config fails
    - Fix tests failure on OpenBSD which comes with Math::BigInt v1.89_01
    - Fix warning on older perls that warned on ref as lvalue in substr

0.05   2015-05-19 22:37:36
    - Statically linked libsodium.a

0.04   2015-05-18 00:31:37
    - BSDs do not come with mempcpy
    - Fix some compiler warnings

0.03   2015-05-12 22:26:26
    - Add Alien::Sodium to configure requirements
    - Correct label for crypto_scalarmult->base

0.02   2015-05-12 01:33:29
    - Add provides for indexing .pod files
    - Minor documentation fixes

0.01   2015-05-12 00:43:48
    - Initial version.



Makefile.PL

use strict;
use warnings;

use ExtUtils::MakeMaker;
use ExtUtils::CBuilder;
use Getopt::Std;
use Config;

my $aliensodium_version = '1.0.7.0';

my %opts;
getopt('L:I:', \%opts) or die usage();

my @defines;
my @dirs;
my $include_dirs;

if ( $opts{L} && $opts{I} ) {
    @dirs = $opts{L};
    $include_dirs = "-I$opts{I}";

    die "Could not locate sodium.h in $opts{I}\n"
        unless -e "$opts{I}/sodium.h";
} else {
    require Alien::Sodium;
    require File::Spec;
    require File::ShareDir;
    require Text::ParseWords;

    my $libsodium = Alien::Sodium->new;

    $include_dirs = $libsodium->cflags;

    @dirs = map { s/^-L//g; $_ } grep { /^-L/ } Text::ParseWords::shellwords($libsodium->libs);

    # NOTE: since Alien::Sodium v1.0.3.1 it may not be needed anymore
    push @dirs, File::Spec->catdir( File::ShareDir::dist_dir("Alien-Sodium"), 'lib' );

    # NOTE: some smokers still cannot find libsodium.a, but don't add it yet, as
    # the updated Alien::Base could have fixed that
    # push @dirs, map { "$_/.libs/" } grep { /_alien/ } @dirs;
}

my $lib_ext = $Config{lib_ext};
my $libsodium_lib = "libsodium${lib_ext}";
my $libsodium_path;

for my $dir ( @dirs ) {
    if ( -e "$dir/$libsodium_lib" ) {
        $libsodium_path = "$dir/$libsodium_lib";
        last;
    }
}

die "Is Alien::Sodium available? Could not locate $libsodium_lib in @dirs\n"
    unless $libsodium_path;

my $cb = ExtUtils::CBuilder->new( quiet => 1 );
push @defines, "AES256GCM_IS_AVAILABLE" if check_if_aes_is_available();

my %WriteMakefileArgs = (
  ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Crypt/NaCl/Sodium.pm', # retrieve abstract from module
       AUTHOR         => "Alex J. G. Burzy\x{144}ski <ajgb\@cpan.org>") : ()),
  "BUILD_REQUIRES" => {
    "ExtUtils::CBuilder" => 0,
  },
  "CONFIGURE_REQUIRES" => {
    "File::ShareDir" => 0,
    "File::Spec" => 0,
    "Alien::Sodium" => $aliensodium_version,
    "ExtUtils::MakeMaker" => "0",
    "ExtUtils::CBuilder" => 0,
  },
  "DISTNAME" => "Crypt-NaCl-Sodium",
  "DEFINE" => join(" ", map { "-D$_" } @defines),
  "EXE_FILES" => [],
  "MYEXTLIB"  => $libsodium_path,
  "INC"       => join(' ', "-I.", $include_dirs),
  "LICENSE" => "perl",
  "NAME" => "Crypt::NaCl::Sodium",
  "PREREQ_PM" => {
    "Alien::Sodium" => $aliensodium_version,
    "Carp" => 0,
    "Sub::Exporter" => 0,
    "XSLoader" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0,
  },
  "VERSION_FROM"     => 'lib/Crypt/NaCl/Sodium.pm',
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Alien::Sodium" => $aliensodium_version,
  "Carp" => 0,
  "Sub::Exporter" => 0,
  "Test::More" => 0,
  "XSLoader" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

sub usage {
    return "Example usage:\n"
        ."\tperl $0 -L /usr/lib/x86_64-linux-gnu/ -I /usr/include/\n"
}

sub check_if_aes_is_available {
    my $test_file_name = "aes-test.c";
    open( my $test_file, ">", $test_file_name) or die("Cannot write $test_file_name: $!");
    print $test_file <<'EOF';
#include <stdio.h>
#include "sodium.h"

int main(int argc, char **argv) {
    int ic, rc;
    ic = sodium_init();
    rc = crypto_aead_aes256gcm_is_available();
    printf("AES256GCM %s available\n", rc ? "is" : "is not");
    exit( rc ? 0 : 1 );
}
EOF

    close($test_file);

    my @include_dirs = map { s/^\s+//; s/\s+$//; $_ } grep { length } split(/\-I/, $include_dirs);

    my $file_obj = eval {
        $cb->compile(
            source => $test_file_name,
            include_dirs => [ '.', @include_dirs ],
        );
    };
    unlink $test_file_name;

    return unless $file_obj;

    my $file_exe = eval {
        $cb->link_executable(
            objects => $file_obj,
            extra_linker_flags => $libsodium_path,
        );
    };
    unlink $file_obj;

    return unless $file_exe;

    my $rv = system("./$file_exe");
    unlink $file_exe;

    return $rv == 0;
}



inc/SimpleProvides.pm
use strict;
use warnings;
package inc::SimpleProvides;

use Moose;
with 'Dist::Zilla::Role::MetaProvider',
    'Dist::Zilla::Role::FileFinderUser' => {
        default_finders => [ ':InstallModules' ],   # this is overridden in dist.ini!
    },
;

sub metadata
{
    my $self = shift;

    my $version = $self->zilla->version;

    return +{
        provides => {
            map {
                # this is an awful hack and assumes ascii package names:
                # please do not cargo-cult this code elsewhere. The proper
                # thing to do is to crack open the file and read the pod name.
                my $filename = $_->name;
                (my $package = $filename) =~ s{[/\\]}{::}g;
                $package =~ s/^lib:://;
                $package =~ s/\.pod$//;
                $package =~ s/\.pm$//;
                $package => { file => $filename, version => $version }
            } @{$self->found_files},
        }
    };
}

__PACKAGE__->meta->make_immutable;


lib/Crypt/NaCl/Sodium.pm
use strict;
use warnings;

#ABSTRACT: NaCl compatible modern, easy-to-use library for  encryption, decryption, signatures, password hashing and more
package Crypt::NaCl::Sodium;

use Carp qw( croak );
use Sub::Exporter;

require XSLoader;
XSLoader::load('Crypt::NaCl::Sodium', $Crypt::NaCl::Sodium::{VERSION} ?
        ${ $Crypt::NaCl::Sodium::{VERSION} } : ()
    );

my @funcs = qw(
    bin2hex hex2bin
    memcmp compare memzero
    increment
    random_bytes
    random_number
);

Sub::Exporter::setup_exporter(
    {
        exports => \@funcs,
        groups => {
            all => \@funcs,
            utils => \@funcs,
        }
    }
);

sub new {
    my ($proto, $submodule) = @_;

    if ( ! $submodule ) {
        my $o = 0;
        return bless \$o, $proto;
    }

    if ( my $m = $proto->can($submodule) ) {
        return $m->();
    }

    croak "Unknown submodule $submodule\n";
}

sub secretbox {
    return Crypt::NaCl::Sodium::secretbox->new();
}

sub auth {
    return Crypt::NaCl::Sodium::auth->new();
}

sub aead {
    return Crypt::NaCl::Sodium::aead->new();
}

sub box {
    return Crypt::NaCl::Sodium::box->new();
}

sub sign {
    return Crypt::NaCl::Sodium::sign->new();
}

sub generichash {
    return Crypt::NaCl::Sodium::generichash->new();
}

sub shorthash {
    return Crypt::NaCl::Sodium::shorthash->new();
}

sub pwhash {
    return Crypt::NaCl::Sodium::pwhash->new();
}

sub hash {
    return Crypt::NaCl::Sodium::hash->new();
}

sub onetimeauth {
    return Crypt::NaCl::Sodium::onetimeauth->new();
}

sub scalarmult {
    return Crypt::NaCl::Sodium::scalarmult->new();
}

sub stream {
    return Crypt::NaCl::Sodium::stream->new();
}

package
    Data::BytesLocker;

our $DEFAULT_LOCKED = 0;

package
    Crypt::NaCl::Sodium::secretbox;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::auth;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::aead;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::box;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::sign;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::generichash;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::shorthash;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::pwhash;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::hash;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::onetimeauth;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::scalarmult;

sub new { return bless {}, __PACKAGE__ }

package
    Crypt::NaCl::Sodium::stream;

sub new { return bless {}, __PACKAGE__ }


1;

__END__

=head1 SYNOPSIS

    use Crypt::NaCl::Sodium qw( :utils );

    my $crypto = Crypt::NaCl::Sodium->new();

    ##########################
    ## Secret-key cryptography

    # Secret-key authenticated encryption (XSalsa20/Poly1305 MAC)
    my $crypto_secretbox = $crypto->secretbox();

    # Secret-key message authentication (HMAC-SHA256, HMAC-SHA512, HMAC-SHA512/256 )
    my $crypto_auth = $crypto->auth();

    # Authenticated Encryption with Additional Data (ChaCha20/Poly1305 MAC, AES256-GCM)
    my $crypto_aead = $crypto->aead();

    ##########################
    ## Public-key cryptography

    # Public-key authenticated encryption (Curve25519/XSalsa20/Poly1305 MAC)
    my $crypto_box = $crypto->box();

    # Public-key signatures (Ed25519)
    my $crypto_sign = $crypto->sign();

    ##########################
    ## Hashing

    # Generic hashing (Blake2b)
    my $crypto_generichash = $crypto->generichash();

    # Short-input hashing (SipHash-2-4)
    my $crypto_shorthash = $crypto->shorthash();

    ##########################
    ## Password hashing (yescrypt)

    my $crypto_pwhash = $crypto->pwhash();

    ##########################
    ## Advanced

    # SHA-2 (SHA-256, SHA-512)
    my $crypto_hash = $crypto->hash();

    # One-time authentication (Poly1305)
    my $crypto_onetimeauth = $crypto->onetimeauth();

    # Diffie-Hellman (Curve25519)
    my $crypto_scalarmult = $crypto->scalarmult();

    # Stream ciphers (XSalsa20, ChaCha20, Salsa20, AES-128-CTR)
    my $crypto_stream = $crypto->stream();

    ##########################
    ## Utilities

    # convert binary data to hexadecimal
    my $hex = bin2hex($bin);

    # convert hexadecimal to binary
    my $bin = hex2bin($hex);

    # constant time comparision of strings
    memcmp($a, $b, $length ) or die '$a ne $b';

    # constant time comparision of large numbers
    compare($x, $y, $length ) == -1 and print '$x < $y';

    # overwrite with null bytes
    memzero($a, $b, ...);

    # generate random number
    my $num = random_number($upper_bound);

    # generate random bytes
    my $bytes = random_bytes($count);

    ##########################
    ## Guarded data storage

    my $locker = Data::BytesLocker->new($password);
    ...
    $locker->unlock();
    print $locker->to_hex();
    $locker->lock();

=head1 DESCRIPTION

L<Crypt::NaCl::Sodium> provides bindings to libsodium - NaCl compatible modern,
easy-to-use library for  encryption, decryption, signatures, password hashing
and more.

It is a portable, cross-compilable, installable, packageable fork
of L<NaCl|http://nacl.cr.yp.to/>, with a compatible API, and an extended API to
improve usability even further.

Its goal is to provide all of the core operations needed to build
higher-level cryptographic tools.

The design choices emphasize security, and "magic constants" have
clear rationales.

And despite the emphasis on high security, primitives are faster
across-the-board than most implementations of the NIST
standards.

L<Crypt::NaCl::Sodium> uses L<Alien::Sodium> that tracks the most current
releases of libsodium.

=head1 METHODS

=head2 new

    my $crypto = Crypt::NaCl::Sodium->new();

Returns a proxy object for methods provided below.

=head2 secretbox

    # Secret-key authenticated encryption (XSalsa20/Poly1305 MAC)
    my $crypto_secretbox = Crypt::NaCl::Sodium->secretbox();

Read L<Crypt::NaCl::Sodium::secretbox> for more details.

=head2 auth

    # Secret-key authentication (HMAC-SHA512/256 and advanced usage of HMAC-SHA-2)
    my $crypto_auth = Crypt::NaCl::Sodium->auth();

Read L<Crypt::NaCl::Sodium::auth> for more details.

=head2 aead

    # Authenticated Encryption with Additional Data (ChaCha20/Poly1305 MAC, AES256-GCM)
    my $crypto_aead = Crypt::NaCl::Sodium->aead();

Read L<Crypt::NaCl::Sodium::aead> for more details.

=head2 box

    # Public-key authenticated encryption (Curve25519/XSalsa20/Poly1305 MAC)
    my $crypto_box = Crypt::NaCl::Sodium->box();

Read L<Crypt::NaCl::Sodium::box> for more details.

=head2 sign

    # Public-key signatures (Ed25519)
    my $crypto_sign = Crypt::NaCl::Sodium->sign();

Read L<Crypt::NaCl::Sodium::sign> for more details.

=head2 generichash

    # Generic hashing (Blake2b)
    my $crypto_generichash = Crypt::NaCl::Sodium->generichash();

Read L<Crypt::NaCl::Sodium::generichash> for more details.

=head2 shorthash

    # Short-input hashing (SipHash-2-4)
    my $crypto_shorthash = Crypt::NaCl::Sodium->shorthash();

Read L<Crypt::NaCl::Sodium::shorthash> for more details.

=head2 pwhash

    # Password hashing (yescrypt)
    my $crypto_pwhash = Crypt::NaCl::Sodium->pwhash();

Read L<Crypt::NaCl::Sodium::pwhash> for more details.

=head2 hash

    # SHA-2 (SHA-256, SHA-512)
    my $crypto_hash = Crypt::NaCl::Sodium->hash();

Read L<Crypt::NaCl::Sodium::hash> for more details.

=head2 onetimeauth

    # One-time authentication (Poly1305)
    my $crypto_onetimeauth = Crypt::NaCl::Sodium->onetimeauth();

Read L<Crypt::NaCl::Sodium::onetimeauth> for more details.

=head2 scalarmult

    # Diffie-Hellman (Curve25519)
    my $crypto_scalarmult = Crypt::NaCl::Sodium->scalarmult();

Read L<Crypt::NaCl::Sodium::scalarmult> for more details.

=head2 stream

    # Stream ciphers (XSalsa20, ChaCha20, Salsa20, AES-128-CTR)
    my $crypto_stream = Crypt::NaCl::Sodium->stream();

Read L<Crypt::NaCl::Sodium::stream> for more details.

=head1 FUNCTIONS

    use Crypt::NaCl::Sodium qw(:utils);

Imports all provided functions.

=head2 bin2hex

    my $hex = bin2hex($bin);

Returns converted C<$bin> into a hexadecimal string.

=head2 hex2bin

    my $hex = "41 : 42 : 43";
    my $bin = hex2bin($hex, ignore => ": ", max_len => 2 );
    print $bin; # AB

Parses a hexadecimal string C<$hex> and converts it to a byte sequence.

Optional arguments:

=over 4

=item * ignore

A string of characters to skip. For example, the string C<": "> allows columns
and spaces to be present at any locations in the hexadecimal string. These
characters will just be ignored.

If unset any non-hexadecimal characters are disallowed.

=item * max_len

The maximum number of bytes to return.

=back

The parser stops when a non-hexadecimal, non-ignored character is
found or when C<max_len> bytes have been written.

=head2 memcmp

    memcmp($a, $b, $length ) or die "\$a ne \$b for length: $length";

Compares strings in constant-time. Returns true if they match, false otherwise.

The argument C<$length> is optional if variables are of the same length. Otherwise it is
required and cannot be greater then the length of the shorter of compared variables.

B<NOTE:> L<Data::BytesLocker/"memcmp"> provides the same functionality.

    $locker->memcmp($b, $length) or die "\$locker ne \$b for length: $length";

=head2 compare

    compare($x, $y, $length ) == -1 and print '$x < $y';

A constant-time version of L</memcmp>, useful to compare nonces and counters
in little-endian format, that plays well with L</increment>.

Returns C<-1> if C<$x> is lower then C<$y>, C<0> if C<$x> and C<$y> are
identical, or C<1> if C<$x> is greater then C<$y>. Both C<$x> and C<$y> are
assumed to be numbers encoded in little-endian format.

The argument C<$length> is optional if variables are of the same length. Otherwise it is
required and cannot be greater then the length of the shorter of compared variables.

B<NOTE:> L<Data::BytesLocker/"compare"> provides the same functionality.

    $locker->compare($y, $length) == -1 and print "\$locker < \$y for length: $length";

=head2 memzero

    memzero($a, $b, ...);

Replaces the value of the provided stringified variables with C<null> bytes. Length of the
zeroed variables is unchanged.

=head2 random_number

    my $num = random_number($upper_bound);

Returns an unpredictable number between 0 and optional C<$upper_bound>
(excluded).
If C<$upper_bound> is not specified the maximum value is C<0xffffffff>
(included).

=head2 increment

    increment($nonce, ...);

B<NOTE:> This function is deprecated and will be removed in next version. Please
use L<Data::BytesLocker/"increment">.

Increments an arbitrary long unsigned number(s) (in place). Function runs in constant-time
for a given length of arguments and considers them to be encoded in
little-endian format.

=head2 random_bytes

    my $bytes = random_bytes($num_of_bytes);

Generates unpredictable sequence of C<$num_of_bytes> bytes.

The length of the C<$bytes> equals the value of C<$num_of_bytes>.

Returns L<Data::BytesLocker> object.

=head1 VARIABLES

=head2 $Data::BytesLocker::DEFAULT_LOCKED

    use Crypt::NaCl::Sodium;
    $Data::BytesLocker::DEFAULT_LOCKED = 1;

By default all values returned from the provided methods are
unlocked L<Data::BytesLocker> objects. If this variable is set to true then
the returned objects are locked and require calling
L<Data::BytesLocker/"unlock"> before accessing.

=head1 SEE ALSO

=over 4

=item * L<Crypt::NaCl::Sodium::secretbox> - Secret-key authenticated encryption (XSalsa20/Poly1305 MAC)

=item * L<Crypt::NaCl::Sodium::auth> - Secret-key message authentication (HMAC-SHA256, HMAC-SHA512, HMAC-SHA512/256 )

=item * L<Crypt::NaCl::Sodium::aead> - Authenticated Encryption with Additional Data (ChaCha20/Poly1305 MAC, AES256-GCM)

=item * L<Crypt::NaCl::Sodium::box> - Public-key authenticated encryption (Curve25519/XSalsa20/Poly1305 MAC)

=item * L<Crypt::NaCl::Sodium::sign> - Public-key signatures (Ed25519)

=item * L<Crypt::NaCl::Sodium::generichash> - Generic hashing (Blake2b)

=item * L<Crypt::NaCl::Sodium::shorthash> - Short-input hashing (SipHash-2-4)

=item * L<Crypt::NaCl::Sodium::pwhash> - Password hashing (yescrypt)

=item * L<Crypt::NaCl::Sodium::hash> - SHA-2 (SHA-256, SHA-512)

=item * L<Crypt::NaCl::Sodium::onetimeauth> - One-time authentication (Poly1305)

=item * L<Crypt::NaCl::Sodium::scalarmult> - Diffie-Hellman (Curve25519)

=item * L<Crypt::NaCl::Sodium::stream> - Stream ciphers (XSalsa20, ChaCha20, Salsa20, AES-128-CTR)

=item * L<Data::BytesLocker> - guarded data storage

=item * L<libsodium|http://jedisct1.gitbooks.io/libsodium> - libsodium

=back

=cut

